### DISCUSSION TITLE
What are your typical navigation flows with Ideavim + IntelliJ ? 😁

### DISCUSSION THREAD

**Discussion by anton-suprun** [JetBrains/ideavim#578] [General]
I'm always looking to improve on my ideavim + IntelliJ workflows and I'm curious what are other people's typical navigation flows?

1. What is your typical approach for vertical navigation?

2. What are your solutions for horizontal navigation?

3. What are some IntelliJ features that you use over vim/ideavim for navigation?

**Comment by anton-suprun** [comment]
Here are mine:

1. Vertical navigation:

- > I use relative line numbers, and jump around with commands like 5j (go down 5 lines) or 7k (go up 7 lines). Biggest disadvantage of this use case is that the jump doesn't get registered in the :jumplist so if you jump and need to come back you either have to mark the spot before you jump or jump back with relative line numbers again. I was hoping that IntelliJ's default Back and Forward action would keep track of relative line jumping but that doesn't work either. Would love to utilize vim's ctrl+I and ctrl+j to jump back and forth between jumplist history but the only way it seems is to use line jumps (I'm referring to the direct 5G, jumps to line 5 and it gets added to jump list but typing out full line numbers is inefficient when there's 3+ digits)

- > Searching the file with vim's / and ? keys. 

- > Bookmarks and mnemonic bookmarks in IntellIJ: I prefer IntelliJ bookmarks over Vim's bookmarks and have set up keybindings in .ideavimrc to set/remove bookmarks, and navigate using bookmarks.

2. Horizontal navigation

- > Using the f character to search. 
- > Tapping w and b
- > Using the $ and ^ symbols to go to start and end of line. 

3. IntelliJ features for navigation:

- > I'm still new to IntelliJ and I recently discovered things like Recent Locations list, the Switcher, recent files. I'm starting to use these features more and more. 
- > Bookmarks and mnemonic bookmarks
- > Jump to previous error and jump to next error actions. I mapped these to a leader key combo and jump up and down the file when addressing any code issues


Would to hear and learn from others' navigation habits!

**Reply by TuralAsgar** [reply]
Jetbrain's not registering to jump back drives me crazy.

**Reply by lippfi** [reply]
> Jetbrain's not registering to jump back drives me crazy.

Could you please describe your issue? I'll try to fix it

**Reply by TuralAsgar** [reply]
@lippfi, thank you for the attention.
@anton-suprun already described.

I use relative line numbers, and jump around with commands like 5j (go down 5 lines) or 7k (go up 7 lines). Biggest disadvantage of this use case is that the jump doesn't get registered in the :jumplist so if you jump and need to come back you either have to mark the spot before you jump or jump back with relative line numbers again. I was hoping that IntelliJ's default Back and Forward action would keep track of relative line jumping but that doesn't work either. Would love to utilize vim's ctrl+I and ctrl+j to jump back and forth between jumplist history but the only way it seems is to use line jumps (I'm referring to the direct 5G, jumps to line 5 and it gets added to jump list but typing out full line numbers is inefficient when there's 3+ digits)

**Reply by lippfi** [reply]
Correct me if I am wrong, but it's expected Vim behavior
You will need something like this, if you want to add some jumps (let's assume more than 5 lines away) to your jump list
```
function! JumpJ()
  if v:count1 > 5
    normal! m`
  endif
  execute 'normal! ' .. v:count1 .. 'j'
endfunction

map j :call JumpJ()<CR>
```
However, I see some issues with this function in IdeaVim:
1. Wrong v:count1 is passed to function
2. Normal command moves caret one char right

I'll address the issues, when I finish some more important fixes

**Reply by anton-suprun** [reply]
@lippfi you're absolutely right. It is expected behaviour in vim. 

I've searched online for an easy workaround via ideavim but haven't found anything yet. I will try what you're suggesting. Thanks!

**Reply by TuralAsgar** [reply]
You are more knowledgeable than me.

I use Neovim.

Whenever I jump let's say 9j (9 lines down) when I press ctrl+o it goes back, and when I click ctrl+i it goes forward.

But in ideavim:
Maybe my problem is different than what anton described.

Cause it doesn't matter 2 lines, or 3 lines the same happens, it doesn't work.

**Reply by anton-suprun** [reply]
@TuralAsgar 

I'm intrigued. I use neovim (mostly) and I don't have that behaviour by default. I wonder if you have some custom re maps for relative line jumps that adds it to the jump list (past locations stored and then accessible via ctrl+i / ctrl+o)

Would you mind sharing your neovim config? I don't think that's the default behaviour in vim/neovim.

**Reply by citizenmatt** [reply]
This has previously been discussed in #661. You can use these mappings:

```vimL
nnoremap <expr> j (v:count <= 1 ? 'j' : "m'" . v:count . 'j')
nnoremap <expr> k (v:count <= 1 ? 'k' : "m'" . v:count . 'k')
xnoremap <expr> j (v:count <= 1 ? 'j' : "m'" . v:count . 'j')
xnoremap <expr> k (v:count <= 1 ? 'k' : "m'" . v:count . 'k')
```

**Reply by TuralAsgar** [reply]
@anton-suprun, of course, but I don't have any. 😁
I am using lazyvim now, and it works.
Even before I used lazyvim, it was working correctly.
Now I am sharing my .ideavimrc, but this was also my .vimrc before.

```
" ------ CONFIG -------

" This makes clipboard and yanking the same
set clipboard+=unnamed
set tabstop=4                   " a tab is four spaces
set softtabstop=4               " when hitting <BS>, pretend like a tab is removed, even if spaces
set shiftwidth=4                " number of spaces to use for autoindenting
set shiftround                  " use multiple of shiftwidth when indenting with '<' and '>'
"set backspace=indent,eol,start  " allow backspacing over everything in insert mode
"set autoindent                  " always set autoindenting on
"set showmatch                   " set show matching parenthesis
"set smarttab                    " insert tabs on the start of a line according to shiftwidth, not tabstop
set expandtab                   " expand tabs to spaces by default
set smartindent
set showmode                    " always show what mode we're currently editing in
set so=999 " Keep cursor centered
set hlsearch
set incsearch " Incrementally search while typing
set ignorecase " Use smart case for searching
set smartcase
set number relativenumber
"make % vim match < > signs
setglobal matchpairs+=<:>

" mute sound
set belloff=all
set visualbell
set t_vb=

" Remaps

" https://www.johnhawthorn.com/2012/09/vi-escape-delays/
" timeoutlen is used for mapping delays, and ttimeoutlen is used for key code delays
set timeoutlen=180 ttimeoutlen=0

" And remember to write the following line before that, to make sure spacebar doesn't have any mapping beforehand:
noremap <SPACE> <Nop>
let mapleader = " "

" delete without yanking
noremap <leader>d "_d
noremap c "_c
noremap C "_C
noremap s "_s
noremap S "_S
"vnoremap <leader>y "+y
"nnoremap <leader>y "+y

" replace currently selected text with default register without yanking it
vnoremap p "_dP

" ------ NAVIGATION

noremap u <Nop>
map J 4j
map K 4k

noremap H _
noremap L $

nnoremap <leader>j J
vnoremap <leader>j J

"write to temporary register t so when another text yanked it will be untouched
"nnoremap <leader>y "ty
"vnoremap <leader>y "ty
"nnoremap <leader>p "tp
"vnoremap <leader>p "tp

"nnoremap <leader>y "+y
"vnoremap <leader>y "+y
"nnoremap <leader>p "+p
"vnoremap <leader>p "+p

nnoremap ' `
nnoremap Y ^vg_y
"nnoremap <leader>n :set invnumber<CR>
inoremap ;; <ESC>A;<ESC>
"inoremap ,, <ESC>A,<ESC>
inoremap ,, <ESC>la, <ESC>
"inoremap >> <ESC>A => ,<ESC>F,i
inoremap >> <ESC>la => ,<ESC>f,i

vnoremap < <gv;
vnoremap > >gv

nnoremap <CR> :nohlsearch<CR>

" set cursor in the center
nnoremap <leader>9 :set so=999<CR>
nnoremap <leader>0 :set so=0<CR>

" move visual code didn't work accordingly so I bound ide action below in ide action section
" Move visual block
"vnoremap J :m '>+1<CR>gv=gv<CR>
"vnoremap K :m '<-2<CR>gv=gv<CR>

" EDITOR ACTIONS
sethandler <C-r> a:ide
sethandler <C-i> a:ide
sethandler <C-o> a:ide
sethandler <C-a> a:vim
sethandler <C-s> a:ide
sethandler <C-x> a:ide
sethandler <C-w> a:ide
sethandler <C-d> a:ide
sethandler <C-u> a:ide
sethandler <C-q> a:ide
sethandler <C-\> a:ide
sethandler <C-v> a:ide
sethandler <C-c> a:ide


set idearefactormode=keep
set ideajoin
set surround
set NERDTree
set ideamarks
set highlightedyank
set easymotion
set keep-english-in-normal

" Use nmap for actions not noremap or nnoremap.
noremap <leader>a :action ToggleBookmark<CR>
noremap <leader>z :action GotoNextBookmark<CR>
noremap <leader>f :action GotoPreviousBookmark<CR>
noremap <leader>b :action ToggleLineBreakpoint<CR>
noremap <leader>e :action QuickEvaluateExpression<CR>
noremap <leader>g :action GithubCopyPathProvider<CR>
noremap <leader>i :action Annotate<CR>
noremap <leader>l :action SelectInProjectView<CR>
noremap <leader>q :action RefactoringMenu<CR>
noremap <leader>r :action RunClass<CR>
noremap <leader>d :action DebugClass<CR>
noremap <leader>hh :action SplitHorizontally<CR>
noremap <leader>h :action MoveTabDown<CR>
noremap <leader>vv :action SplitVertically<CR>
noremap <leader>v :action MoveTabRight<CR>
noremap <leader>c :action MaximizeEditorInSplit<CR>
noremap <leader>t :action PrevSplitter<CR>
noremap <leader>y :action NextSplitter<CR>
noremap <leader>m :action VcsShowPrevChangeMarker<CR>
noremap <leader>n :action VcsShowNextChangeMarker<CR>
noremap <leader>o :action MoveEditorToOppositeTabGroup<CR>
nnoremap <leader>w :action Unwrap<CR>
vnoremap <leader>w :action SurroundWith<CR>
nnoremap <leader><leader> mZ :action FindUsages<CR>
noremap U :action UnsplitAll<CR>
noremap u :action ExpandToLevel1<CR>
noremap [ :action MethodDown<CR>
noremap ] :action MethodUp<CR>
"multiple action mapping 👇
"":action ReformatCode  \| action RiderUnitTestRunSolutionAction<CR>

" Harpoon keybindings
" Setting harpoons
nnoremap <leader><leader>1 :action SetHarpoon1<CR>
nnoremap <leader><leader>2 :action SetHarpoon2<CR>
nnoremap <leader><leader>3 :action SetHarpoon3<CR>
nnoremap <leader><leader>4 :action SetHarpoon4<CR>
nnoremap <leader><leader>5 :action SetHarpoon5<CR>
nnoremap <leader>u :action ShowHarpoon<CR>

" Navigationing harpoons
noremap <leader>1 :action GotoHarpoon1<CR>
noremap <leader>2 :action GotoHarpoon2<CR>
noremap <leader>3 :action GotoHarpoon3<CR>
noremap <leader>4 :action GotoHarpoon4<CR>
noremap <leader>5 :action GotoHarpoon5<CR>
"END Harpoon keybindings

" Flutter remaps
"noremap <leader>w :action Flutter.ExtractWidget<CR>

"mark as a global mark/
nnoremap ma mA
nnoremap mb mB
nnoremap mc mC
nnoremap md mD
nnoremap me mE
nnoremap mf mG
nnoremap mh mH
nnoremap mi mI
nnoremap mj mJ
nnoremap mk mK
nnoremap ml mL
nnoremap mm mM
nnoremap mn mN
nnoremap mo mO
nnoremap mp mP
nnoremap mq mQ
nnoremap mr mR
nnoremap ms mS
nnoremap mt mT
nnoremap mu mU
nnoremap mv mV
nnoremap mw mW
nnoremap mx mX
nnoremap my mY
nnoremap mz mZ

"go to global marks
nnoremap 'a `A
nnoremap 'b `B
nnoremap 'c `C
nnoremap 'd `D
nnoremap 'e `E
nnoremap 'f `G
nnoremap 'h `H
nnoremap 'i `I
nnoremap 'j `J
nnoremap 'k `K
nnoremap 'l `L
nnoremap 'm `M
nnoremap 'n `N
nnoremap 'o `O
nnoremap 'p `P
nnoremap 'q `Q
nnoremap 'r `R
nnoremap 's `S
nnoremap 't `T
nnoremap 'u `U
nnoremap 'v `V
nnoremap 'w `W
nnoremap 'x `X
nnoremap 'y `Y
nnoremap 'z `Z

```

**Reply by TuralAsgar** [reply]
@citizenmatt , man how can I thank you or serve you for this help?
I have been looking for that for long, and none of the solutions worked.
I am not that knowledgeable about config stuff, but I am so addicted to working with i3, yabai, ideavim, shortcuts and anything that helps me do things with the keyboard fast.

This was one of the features I was missing from Neovim.
I love JetBrains IDEs.

Thank you very much!
I am so grateful!

**Reply by citizenmatt** [reply]
Glad it's working! But you need to thank @lippfi for implementing this 😁

**Reply by TuralAsgar** [reply]
@lippfi, thank you very much! 🙏 🙂

**Reply by HosseyNJF** [reply]
@TuralAsgar Just wanted to say that 2 years later, your .ideavimrc is giving me great insights on what's possible using IdeaVim. Thanks for sharing!

**Reply by TuralAsgar** [reply]
> @TuralAsgar Just wanted to say that 2 years later, your .ideavimrc is giving me great insights on what's possible using IdeaVim. Thanks for sharing!

Salam, Hüseyn. Təşəkkür edirəm.
I am really happy to know it.
Thank you for taking your time to let me know and make my day.
I added you to my linkedin, we will talk there. 🙏🏼
